plugins/suggested-blocks/src/index.js

```diff
@@ -0,0 +1,204 @@
+/**
+ * @license
+ * Copyright 2012 Google LLC
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * @fileoverview Utility functions for handling suggestions.
+ */
+'use strict';
+
+/**
+ * Utility functions for handling block suggestions.
+ * @namespace Blockly.SuggestedBlocks
+ */
+import * as Blockly from 'blockly';
+
+const NUM_BLOCKS_PER_CATEGORY = 10;
+
+// Map from workspace ID to BlockSuggestor objects
+const suggestorLookup = new WeakMap();
+
+/**
+ *
+ */
+export class BlockSuggestor {
+  /**
+   * Constructs a BlockSuggestor object.
+   */
+  constructor() {
+    this.defaultJsonForBlockLookup = {};
+    this.recentlyUsedBlocks = [];
+    this.workspaceHasFinishedLoading = false;
+
+    this.eventListener = this.eventListener.bind(this);
+    this.getMostUsed = this.getMostUsed.bind(this);
+    this.getRecentlyUsed = this.getRecentlyUsed.bind(this);
+  }
+
+  /**
+   * Generates a list of the 10 most frequently used blocks, in order.
+   * Includes a secondary sort by most recent blocks.
+   * @return {Array <object>}A list of block JSON objects
+   */
+  getMostUsed = function() {
+    // Store the frequency of each block, as well as the index first appears at
+    const countMap = {};
+    const recencyMap = {};
+    for (const [index, key] of this.recentlyUsedBlocks.entries()) {
+      countMap[key] = (countMap[key] || 0) + 1;
+      if (!recencyMap[key]) {
+        recencyMap[key] = index + 1;
+      }
+    }
+
+    // Get a sorted list
+    const freqUsedBlockTypes = [];
+    for (const key of Object.keys(countMap)) {
+      freqUsedBlockTypes.push(key);
+    }
+    freqUsedBlockTypes.sort((a, b) => countMap[b] - countMap[a] +
+        0.01 * (recencyMap[a] - recencyMap[b]));
+
+    return this.generateBlockData(freqUsedBlockTypes);
+  };
+
+  /**
+   * Generates a list of the 10 most recently used blocks.
+   * @return {Array <object>} A list of block JSON objects
+   */
+  getRecentlyUsed = function() {
+    const uniqueRecentBlocks = [...new Set(this.recentlyUsedBlocks)];
+    const recencyMap = {};
+    for (const [index, key] of this.recentlyUsedBlocks.entries()) {
+      if (!recencyMap[key]) {
+        recencyMap[key] = index + 1;
+      }
+    }
+    uniqueRecentBlocks.sort((a, b) => recencyMap[a] - recencyMap[b]);
+    return this.generateBlockData(uniqueRecentBlocks);
+  }
+
+  generateBlockData = function(blockTypeList) {
+    const blockList = [];
+    for (const key of blockTypeList.slice(0, NUM_BLOCKS_PER_CATEGORY)) {
```

Comment:
If you want I think you could use `blockTypeList.slice(0, num).map()`. Not sure about your feelings on mixing pure code with imperative code :P